### DISCUSSION TITLE
Problem installing python-kasa getting 'Building wheel for maturin (pyproject.toml) did not run successfully'

### DISCUSSION THREAD

**Discussion by as2651** [termux/termux-app#3860] [Q&A]
I have python v3.11 with pip v24 installed. Doing 'pip install python-kasa', and getting following error stack.
No issues with installing and using python-kasa on Pydroid3 app.
Any ideas what's wrong with Termux?

error: subprocess-exited-with-error

× pip subprocess to install build dependencies did not run successfully.
│ exit code: 1
╰─> [60 lines of output]
Collecting maturin<2,>=1
Using cached maturin-1.4.0.tar.gz (177 kB)
Installing build dependencies: started
Installing build dependencies: finished with status 'done'
Getting requirements to build wheel: started
Getting requirements to build wheel: finished with status 'done'
Preparing metadata (pyproject.toml): started
Preparing metadata (pyproject.toml): finished with status 'done'
Collecting typing-extensions!=4.7.0,>=4.6.0
Using cached typing_extensions-4.10.0-py3-none-any.whl.metadata (3.0 kB)
Using cached typing_extensions-4.10.0-py3-none-any.whl (33 kB)
Building wheels for collected packages: maturin
Building wheel for maturin (pyproject.toml): started
Building wheel for maturin (pyproject.toml): finished with status 'error'
error: subprocess-exited-with-error

    × Building wheel for maturin (pyproject.toml) did not run successfully.
    │ exit code: 1
    ╰─> [35 lines of output]
        /data/data/com.termux/files/usr/tmp/pip-build-env-rq7v82ra/overlay/lib/python3.11/site-packages/setuptools/config/_apply_pyprojecttoml.py:83: SetuptoolsWarning: `install_requires` overwritten in `pyproject.toml` (dependencies)
          corresp(dist, value, root_dir)
        running bdist_wheel
        running build
        running build_py
        creating build
        creating build/lib.linux-aarch64-cpython-311
        creating build/lib.linux-aarch64-cpython-311/maturin
        copying maturin/__init__.py -> build/lib.linux-aarch64-cpython-311/maturin
        copying maturin/__main__.py -> build/lib.linux-aarch64-cpython-311/maturin
        copying maturin/import_hook.py -> build/lib.linux-aarch64-cpython-311/maturin
        running egg_info
        creating maturin.egg-info
        writing maturin.egg-info/PKG-INFO
        writing dependency_links to maturin.egg-info/dependency_links.txt
        writing requirements to maturin.egg-info/requires.txt
        writing top-level names to maturin.egg-info/top_level.txt
        writing manifest file 'maturin.egg-info/SOURCES.txt'
        reading manifest file 'maturin.egg-info/SOURCES.txt'
        reading manifest template 'MANIFEST.in'
        warning: no files found matching '*.json' under directory 'src/python_interpreter'
        writing manifest file 'maturin.egg-info/SOURCES.txt'
        running build_ext
        running build_rust
        error: can't find Rust compiler

        If you are using an outdated pip version, it is possible a prebuilt wheel is available for this package but pip is not able to install from it. Installing from the wheel would avoid the need for a Rust compiler.

        To update pip, run:

            pip install --upgrade pip

        and then retry package installation.

        If you did intend to build this package from source, try installing a Rust compiler from your system package manager and ensure it is on the PATH during installation. Alternatively, rustup (available at https://rustup.rs/) is the recommended way to download and update the Rust compiler toolchain.
        [end of output]

    note: This error originates from a subprocess, and is likely not a problem with pip.
    ERROR: Failed building wheel for maturin
  Failed to build maturin
  ERROR: Could not build wheels for maturin, which is required to install pyproject.toml-based projects
  [end of output]
note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× pip subprocess to install build dependencies did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.

**Comment by Biswa96** [comment]
Did you install rust?

**Reply by as2651** [reply]
python-kasa folks are saying that it shouldn't be required

**Reply by Biswa96** [reply]
I have tried the `pip install python-kasa` command in clean termux install. It failed with same errors as yours. The dependency goes like python-kasa > pydantic > pydantic-core > maturin.

**Comment by cujoXavier** [comment]
I tried all the same things as @rytilahti including installing `rust`

and still, no wheels for `maturin` -- same error and flags.

then I tried installing `rustup` for use as a rust compliler via `wget -i` <https://rustup.rs>

this is when I noticed this was calling for a tool from `aarch64-linux-android-ar`

and that was the key...

I searched github & found the answer:

the GitHub link below 👇🏽 shows that it is as simple as 
`pkg install binutils`

and so I tried that, followed by again trying to run `pip install python-kasa`

 And it worked perfectly ^_~ 👍🏽

<https://github.com/termux/termux-packages/discussions/8214>

**Reply by cujoXavier** [reply]
I updated my answer to remove all references to the 
person @rytilahti -- who previously answered this Q&A saying, "this wasn't an issue with pip or anything with the kasa build, but rather a bug with Termux and suggested @as2651 should send it to the termux development team. And then `closed` the issue as `resolved.`

However, since I called them out on that directly, and provided a very simple solution-  available _right here_ in [termux/termux-packages #8214](url)

@rytilahti deleted their comments and `reverted` the status from `closed` to `none yet.`

Therefore, I have edited my answer to remove all references to this person.

I'm just happy that there was a straightforward answer to this issue.

@as2651 please let me know if you were also able to get the installation done or if you need further assistance.

If you have had success, this can now be closed as resolved.

Best of luck.

-CujoX

**"never duck and cover, it just makes you look even weaker than you already are"**

**Reply by ALIYUsbly** [reply]
This was really helpful I have traied many solutions only this one work thanks.
As for me I was trying to "pip install assemblyai" which has pydentic-core as dependency that's was when "can't find Rust compiler"  error rised
